src/target/riscv/riscv-013_reg.c

```diff
@@ -13,73 +13,454 @@
+			buf_get_u32(priv_buf, VIRT_PRIV_PRV_OFFSET, VIRT_PRIV_PRV_LENGTH));
+	buf_set_u32(dcsr_buf, CSR_DCSR_V_OFFSET, CSR_DCSR_V_LENGTH,
+			buf_get_u32(priv_buf, VIRT_PRIV_V_OFFSET, VIRT_PRIV_V_LENGTH));
+	res = dcsr->type->set(dcsr, dcsr_buf);
+	if (res == ERROR_OK)
+		set_cache_value(priv, priv_buf);
+	return res;
+}
+
+/* End of "reg_arch_type" method definitions. */
+
+static const struct reg_arch_type zero_type = {
+	.get = riscv013_caching_get,
+	/* "zero" is read-only, but there is nothig wrong with attempting to
+	 * write it (e.g. for the purpose of HW testing). */
+	.set = riscv013_noncaching_set,
+	.flush = riscv013_reg_flush,
+};
+
+static const struct reg_arch_type xreg_type = {
+	.get = riscv013_caching_get,
+	.set = riscv013_caching_set,
+	.flush = riscv013_reg_flush,
+};
+
+static const struct reg_arch_type freg_type = {
+	.get = riscv013_caching_get,
+	.set = riscv013_caching_set,
+	.flush = riscv013_reg_flush,
+};
+
+static const struct reg_arch_type vreg_type = {
+	.get = get_vreg,
+	.set = set_vreg,
+	.flush = flush_vreg
+};
+
+static const struct reg_arch_type pc_type = {
+	.get = get_pc,
+	.set = set_pc,
+	.flush = flush_noncacheable
+};
+
+static const struct reg_arch_type priv_type = {
+	.get = get_priv,
+	.set = set_priv,
+	.flush = flush_noncacheable
+};
+
+static const struct reg_arch_type warl_csr_type = {
+	.get = riscv013_caching_get,
+	.set = riscv013_noncaching_set,
+	.flush = riscv013_reg_flush
+};
+
+static const struct reg_arch_type noncacheable_type = {
+	.get = riscv013_noncaching_get,
+	.set = riscv013_noncaching_set,
+	.flush = flush_noncacheable
+};
+
 static const struct reg_arch_type *riscv013_gdb_regno_reg_type(uint32_t regno)
 {
-	static const struct reg_arch_type riscv013_reg_type = {
-		.get = riscv013_reg_get,
-		.set = riscv013_reg_set
-	};
-	return &riscv013_reg_type;
+	if (regno != GDB_REGNO_ZERO && regno <= GDB_REGNO_XPR31)
+		return &xreg_type;
+	if (regno >= GDB_REGNO_FPR0 && regno <= GDB_REGNO_FPR31) {
+		/* For now "freg_type" is the same as "xreg_type", but it will be
+		 * changed soon */
+		return &freg_type;
+	}
+	if (regno >= GDB_REGNO_V0 && regno <= GDB_REGNO_V31)
+		return &vreg_type;
+	switch (regno) {
+	case GDB_REGNO_ZERO:
+		return &zero_type;
+	case GDB_REGNO_PC:
+		return &pc_type;
+	case GDB_REGNO_PRIV:
+		return &priv_type;
+	case GDB_REGNO_VLENB:
+	/* "vlenb" is read-only, but there is nothig wrong with attempting to
+	 * write it (e.g. for the purpose of HW testing). */
+	case GDB_REGNO_DPC:
+	case GDB_REGNO_VSTART:
+	case GDB_REGNO_VXSAT:
+	case GDB_REGNO_VXRM:
+	case GDB_REGNO_VL:
+	case GDB_REGNO_VTYPE:
+	case GDB_REGNO_MISA:
+	case GDB_REGNO_DCSR:
+	case GDB_REGNO_DSCRATCH0:
+	case GDB_REGNO_MEPC:
+	case GDB_REGNO_SATP:
+		return &warl_csr_type;
+	/* TODO: https://github.com/riscv-collab/riscv-openocd/issues/1219 */
+	case GDB_REGNO_MSTATUS:
+	case GDB_REGNO_MCAUSE:
+	/* TODO: Sdtrig register should be WARL. */
```

Comment:
IIUC, any cached values of `tdata*` and `tinfo` must get invalidated when `tselect` changes.

This can be mentioned in the TODO:

```suggestion
	/* TODO: Sdtrig register should be WARL. 
	 * Also cached values of  of "tdata*" and "tinfo" must get invalidated when tselect is changed. */
```

Reply:
Addressed.